core/image_io/png.cpp

```diff
@@ -59,19 +56,15 @@ namespace MR
 
     void PNG::unload (const Header& header)
     {
-      if (addresses.size()) {
-        if (writable) {
-          size_t slice_bytes = (header.datatype().bits() * header.size(0) * header.size(1) + 7) / 8;
-          if (header.ndim() == 4)
-            slice_bytes *= header.size (3);
-          for (size_t i = 0; i != files.size(); i++) {
-            File::PNG::Writer png (header, files[i].name);
-            png.save (addresses[0].get() + (i * slice_bytes));
-          }
+      assert (addresses.size() == 1);
+      if (writable) {
+        const size_t slice_bytes = (header.datatype().bits() * header.size(0) * header.size(1) * (header.ndim() == 4 ? header.size(3) : 1) + 7) / 8;
+        for (size_t i = 0; i != files.size(); i++) {
+          File::PNG::Writer png (header, files[i].name);
+          png.save (addresses[0].get() + (i * slice_bytes));
         }
-        DEBUG ("deleting buffer for PNG image \"" + header.name() + "\"...");
-        addresses[0].release();
       }
+      delete[] addresses[0].release();
```

Comment:
This is necessary because `MR::ImageIO::Base::addresses` is of type `vector<std::unique_ptr<uint8_t[]>>. However there are other classes in `MR::ImageIO` that do *not* invoke `delete[]`. My suspicion is that in this PR I caught a memory leak when diagnosing other PNG issues, but this is actually a more general memory leak / issue across may image formats:

-   `Default` just calls `addresses[n].release()`
-   `GZ` does nothing
-   `Mosaic` does nothing
-   `Pipe` just calls `addresses[0].release()`
-   `RAM` calls `delete [] addresses[0]`
-   `Scratch` calls `addresses[0].reset()`
-   `TIFF` calls `addresses[0].release()`
-   `VariableScaling` does nothing

Reply:
I think it's enough for that line to just disappear - the `addresses` vector will be cleared immediately afterwards anyway, and the default destructor for `std::unique_ptr<uint8_t[]>` will free the associated memory anyway.